### PR TITLE
[pm] E: Isolate Thread Data Structure from Process

### PR DESCRIPTION
**Pull request by ravixr**
This PR is an initial solution to the issue described at #383.

### CODE REVIEW

File: src/kernel/kcall/thread.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *============================================================================*/
+
+/**
+ * @details Creates a new thread.
+ */
+tid_t kcall_thread_create(pid_t pid)
+{
+    // Check if pid is valid.
+    if (!WITHIN(pid, 0, PROCESS_MAX))
+        return (-EINVAL);
```
Comment (ppenna) [inline src/kernel/kcall/thread.c:25]: Please, enclose this in a `{}`

File: src/kernel/kcall/thread.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    UNREACHABLE();
+}
+
+/**
+ * @details Returns the ID of the calling thread.
+ */
+tid_t kcall_thread_get_id(void)
+{
+    return (thread_get_curr()->tid);
+}
```
Comment (ppenna) [inline src/kernel/kcall/thread.c:55]: Please add an empty line on the end of this file.

File: src/kernel/pm/thread.c [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    return (NULL);
+}
+
+/**
+ * @details Releases a thread entry.
+ */
+int thread_free(pid_t pid)
+{
+    if (pid <= KERNEL_PROCESS || pid >= PROCESS_MAX)
+        return (-EINVAL);
```
Comment (ppenna) [inline src/kernel/pm/thread.c:135]: Please enclose this in `{}`.

File: src/kernel/pm/thread.c [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    return (running);
+}
+
+/**
+ * @details Checks if a TID refers to a valid thread.
+ */
+struct thread *thread_get_tcb(tid_t tid)
+{
+    if (tid < 0 || tid >= THREADS_MAX)
+        return (NULL);
```
Comment (ppenna) [inline src/kernel/pm/thread.c:162]: Please enclose this in `{}`.

File: src/kernel/pm/thread.c [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ * @details This function terminates the calling thread.
+ */
+noreturn void thread_exit(void)
+{
+    thread_free(running->pid);
+    thread_yield();
+    UNREACHABLE();
+}
```
Comment (ppenna) [inline src/kernel/pm/thread.c:236]: Please add an empty line in the end of this file.

File: src/kernel/pm/thread.c [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ * @details This function wakes up all threads owned by the process pointed to
+ * by @p pid.
+ */
+void thread_wakeup(pid_t pid)
+{
+    if (pid < 0 || pid >= PROCESS_MAX)
+        return;
```
Comment (ppenna) [inline src/kernel/pm/thread.c:219]: Please enclose this in `{}`

File: src/kernel/pm/thread.c [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+ * @details This function puts all threads owned by the calling process to
+ * sleep. The calling process resumes its execution when another process invokes
+ * `process_wakeup()` which internally invokes `thread_wakeup()`.
+ */
+void thread_sleep(void)
+{
+    for (int i = 0; i < THREADS_MAX; i++) {
+        if (threads[i].pid == running->pid) {
+            threads[i].state = THREAD_WAITING;
+        }
```
Comment (ppenna) [inline src/kernel/pm/thread.c:206]: What if the thread was terminated, then will you place it in the waiting queue? Note that you don't unassigned the PID when freeing the thread.

File: src/kernel/pm/thread.c [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    if (pid <= KERNEL_PROCESS || pid >= PROCESS_MAX)
+        return (-EINVAL);
+
+    for (int i = 0; i < THREADS_MAX; i++) {
+        if (threads[i].pid == pid) {
+            KASSERT(kpage_put(threads[i].stack) == 0);
+            threads[i].stack = NULL;
+            threads[i].state = THREAD_AVAIABLE;
+        }
+    }
```
Comment (ppenna) [inline src/kernel/pm/thread.c:143]: Shouldn't you unassigned the PID?

File: include/nanvix/kernel/kcall.h [hunk 9]
```diff
@@ -31,52 +31,57 @@
...
-#define NR_upage_alloc    /** kernel_upage_alloc()      */
-#define NR_upage_free     /** kernel_upage_free()       */
-#define NR_upage_map      /** kernel_upage_map()        */
-#define NR_upage_link     /** kernel_upage_link()       */
-#define NR_upage_unlink   /** kernel_upage_unlink()     */
-#define NR_upage_unmap    /** kernel_upage_unmap()      */
-#define NR_excp_ctrl      /** kernel_excp_ctrl()        */
-#define NR_excp_pause     /** kernel_excp_pause()       */
-#define NR_excp_resume    /** kernel_excp_resume()      */
+#define NR_void0 0          /** kernel_void0()            */
```
[the remark this answers is not in the document]
Comment (ravixr) [reply include/nanvix/kernel/kcall.h:34]: I believe the added definitions made the formatting shift all the doc comments.

[the remark this answers is not in the document]
Comment (ppenna) [reply include/nanvix/kernel/kcall.h:34]: The order of items has also changed. Why?

**Review by ppenna — APPROVED** [review]
LGTM

**Review by ppenna — APPROVED** [review]

**Pull request merged**